src/consensus/malachite/read_host.rs

```diff
@@ -0,0 +1,146 @@
+//! Implementation of a host actor for bridiging consensus and the application via a set of channels.
+
+use crate::core::types::SnapchainValidatorContext;
+use crate::proto::{self, Height};
+use crate::storage::store::engine::{BlockEngine, ShardEngine};
+use bytes::Bytes;
+use informalsystems_malachitebft_sync::DecidedValue;
+use prost::Message;
+use ractor::{async_trait, Actor, ActorProcessingErr, ActorRef, RpcReplyPort, SpawnErr};
+use tracing::error;
+
+/// Messages that need to be handled by the host actor.
+#[derive(Debug)]
+pub enum ReadHostMsg {
+    /// Request the earliest block height in the block store
+    GetHistoryMinHeight {
+        reply_to: RpcReplyPort<Height>,
+    },
+
+    // Consensus has decided on a value
+    ProcessDecidedValue {
+        value: proto::DecidedValue,
+    },
+
+    // Retrieve decided block from the block store
+    GetDecidedValue {
+        height: Height,
+        reply_to: RpcReplyPort<Option<DecidedValue<SnapchainValidatorContext>>>,
+    },
+}
+
+pub type ReadHostRef = ActorRef<ReadHostMsg>;
+
+pub struct ReadHost {}
+
+pub enum Engine {
+    ShardEngine(ShardEngine),
+    BlockEngine(BlockEngine),
+}
+
+pub struct ReadHostState {
+    pub engine: Engine,
+    pub shard_id: u32,
+}
+
+impl ReadHost {
+    pub fn new() -> Self {
+        ReadHost {}
+    }
+
+    pub async fn spawn(state: ReadHostState) -> Result<ActorRef<ReadHostMsg>, SpawnErr> {
+        let (actor_ref, _) = Actor::spawn(None, Self::new(), state).await?;
+        Ok(actor_ref)
+    }
+}
+
+impl ReadHost {
+    async fn handle_msg(
+        &self,
+        _myself: ActorRef<ReadHostMsg>,
+        msg: ReadHostMsg,
+        state: &mut ReadHostState,
+    ) -> Result<(), ActorProcessingErr> {
+        match msg {
+            ReadHostMsg::GetHistoryMinHeight { reply_to } => {
+                reply_to.send(crate::proto::Height::new(state.shard_id, 1))?;
+            }
+
+            ReadHostMsg::ProcessDecidedValue { value } => match &mut state.engine {
```

Comment:
I think we'll need to have some state here to handle missing heights and non-consecutive heights. The ordering of the messages here is not guaranteed. 

Also, we should add an explicit check within `commit_...` to check that we're replaying the very next height not not anything else (maybe within `replay_proposal`?)